File: CombinedAlgorithms.java
import java.util.Random;

public class CombinedAlgorithms {

    // Sum Array algorithm
    public static int sumArrayMembers(int[] arr) {
        double startTime = System.nanoTime(); // Start the timer

        // Sum Array algorithm
        int sum = 0;
        for (int i = 0; i < arr.length; i++) {
            sum += arr[i];
        }

        double endTime = System.nanoTime(); // End the timer
        // Nanoseconds to seconds
        double duration = (endTime - startTime);
        System.out.println("Execution time: " + duration + " ns");

        // Nanoseconds to seconds
        double timeSeconds = duration / 1000000000;
        System.out.println("Execution time: " + String.format("%,.9f", timeSeconds) + " seconds or " + String.format("%,.1f", duration) + " nanoseconds");

        return sum;
    }

    // scalar_multiplication algorithm
    public static void scalar_multiplication(int[] arr, int scalar) {
        int row;
        int column;
        int randomrange = arr.length;

        row = (int) (Math.random() * randomrange);
        column = (int) (Math.random() * randomrange);

        //create empty matrix's given size
        int[][] A = new int[row][column];
        int[][] C = new int[row][column];

        for (int i = 0; i < row; i++) {
            for (int j = 0; j < column; j++) {
                A[i][j] = (int) (Math.random() * randomrange);
            }
        }

        double startTime = System.nanoTime(); // Start the timer

        // Perform scalar multiplication
        for (int i = 0; i < row; i++) {
            for (int j = 0; j < column; j++) {
                C[i][j] = A[i][j] * scalar;
            }
        }

        double endTime = System.nanoTime(); // End the timer

        double duration = (endTime - startTime);
        // Nanoseconds to seconds
        double timeSeconds = duration / 1000000000;

//        // Print old matrix
//        System.out.println("Old Matrix: ");
//        for (int i = 0; i < row; i++) {
//            for (int j = 0; j < column; j++) {
//                System.out.print(A[i][j] + " ");
//            }
//            System.out.println();
//        }
//
//        // Print new matrix
//        System.out.println("New Matrix: ");
//        for (int i = 0; i < row; i++) {
//            for (int j = 0; j < column; j++) {
//                System.out.print(C[i][j] + " ");
//            }
//            System.out.println();
//        }


        //Print statements with input size and execution time
        System.out.println("Input Size: " + row + "x" + column);
        System.out.println("Execution time: " + String.format("%,.9f", timeSeconds) + " seconds or " + String.format("%,.1f", duration) + " nanoseconds");
        System.out.println("Scalar multiplier is " + scalar);

    }

    //Method to generate a matrix with numbers 0-9 and a size of n x n
    public static int[][] generateMatrix(int n)
    {
        Random random = new Random();
        int[][] matrix = new int[n][n];
        for (int i = 0; i < n; i++)
        {
            for (int j = 0; j < n; j++)
            {
                matrix[i][j] = random.nextInt(10);
            }
        }
        return matrix;
    }

    //Performs matrix multiplication for matrix A and B
    public static void matrixMultiplication(int n, int[][] A, int[][] B)
    {
        //Initialize result matrix
        int[][] C = new int[n][n];

        //Start timer measured in nanoseconds
        double startTime = System.nanoTime();

        for (int i = 0; i < n; i++)
        {
            for (int j = 0; j < n; j++)
            {
                for (int k = 0; k < n; k++)
                {
                    C[i][j] += A[i][k] * B[k][j];
                }
            }
        }

        //Stop timer
        double endTime = System.nanoTime();

        //Calculate time in Nanoseconds and Seconds
        double timeNano = endTime - startTime;
        // Nanoseconds to seconds
        double timeSeconds = timeNano / 1000000000;

        //Print statements with input size and execution time
        System.out.println("Input Size: " + n);
        System.out.println("Execution time: " + String.format("%,.9f", timeSeconds) + " seconds or " + String.format("%,.1f", timeNano) + " nanoseconds");

    }

    public static void main(String[] args) {
        // Test each algorithm with 3 different cases (n = 100, 1000, 10000)
        int[] arr1 = new int[100];
        int[] arr2 = new int[1000];
        int[] arr3 = new int[10000];
        for (int i = 0; i < arr1.length; i++) {
            arr1[i] = i + 1;
        }
        for (int i = 0; i < arr2.length; i++) {
            arr2[i] = i + 1;
        }
        for (int i = 0; i < arr3.length; i++) {
            arr3[i] = i + 1;
        }

        // Test algorithmOne
        System.out.println("Algorithm 1");
        System.out.println("Size of array: " + arr1.length);
        System.out.println(sumArrayMembers(arr1));
        System.out.println("Size of array: " + arr2.length);
        System.out.println(sumArrayMembers(arr2));
        System.out.println("Size of array: " + arr3.length);
        System.out.println(sumArrayMembers(arr3));

        // Test scalar_multiplication
        System.out.println("Scalar Multiplication");
        scalar_multiplication(arr1, 5);
        scalar_multiplication(arr2, 5);
        scalar_multiplication(arr3, 5);

        // Test matrixMultiplication
        System.out.println("Matrix Multiplication");
        int[][] matrix1 = generateMatrix(100);
        int[][] matrix2 = generateMatrix(100);
        int[][] matrix3 = generateMatrix(100);
        matrixMultiplication(10, matrix1, matrix2);
        matrixMultiplication(50, matrix2, matrix3);
        matrixMultiplication(100, matrix3, matrix1);


    }

}

File: CombinedAlgorithms.py

import time
import random

# Function to sum all elements in an array
def sumArray(inputArray):
    # Start timer for algorithm
    start_time = time.perf_counter_ns()

    sum = 0
    # For function to sum all elements in array
    for i in range(len(inputArray)):
        sum = sum + inputArray[i]

    # End timer for algorithm (All elements have been summed)
    end_time = time.perf_counter_ns()
    # Calculate the time it took
    duration_ns = (end_time - start_time)

    timeTook = duration_ns

    # nanoseconds to seconds
    timeTookSeconds = timeTook / 1e9

    print("Input Size: ", len(inputArray))
    print("Execution time: ", timeTook, " nanoseconds ", "(", timeTookSeconds, " seconds )")

    return sum, timeTook

# Scalar Multiplication

def scalar_multiplication(arr, scalar):
    row = int(random.uniform(0, arr))
    column = int(random.uniform(0, arr))

    # fill empty matrix A with 0
    A = [[0 for _ in range(column)] for _ in range(row)]
    C = [[0 for _ in range(column)] for _ in range(row)]

    for i in range(row):
        for j in range(column):
            A[i][j] = int(random.uniform(0, arr))

    scalar = 5

    # start time
    start_time = time.perf_counter_ns()

    # scalar matrix multiplication

    for i in range(row):
        for j in range(column):
            C[i][j] = A[i][j] * scalar

    # end time
    end_time = time.perf_counter_ns()
    duration = (end_time - start_time)

    timeTook = duration

    # nanoseconds to seconds
    timeTookSeconds = timeTook / 1e9

    print("Input Size: ", row, "x", column)
    print("Execution time: ", timeTook, " nanoseconds ", "(", timeTookSeconds, " seconds )")
    print("Scalar multiplier is ", scalar)

    return C, timeTook

# Matrix Multiplication

#Method to generate a matrix with numbers from 0-9 and a size of n x n
def generate_matrix(n):
    return [[random.randint(0, 9) for _ in range(n)] for _ in range(n)]


# Performs matrix multiplication for matrix A and B
def matrix_multiplication(n, A, B):
    # Fill result matrix with zeros
    C = [[0 for _ in range(n)] for _ in range(n)]

    # Start timer measured in nanoseconds
    start_time = time.perf_counter_ns()

    for i in range(n):
        for j in range(n):
            for k in range(n):
                C[i][j] += A[i][k] * B[k][j]

    # Stop timer
    end_time = time.perf_counter_ns()

    # Calculate the time it took
    duration_ns = (end_time - start_time)

    timeTook = duration_ns

    # nanoseconds to seconds
    timeTookSeconds = timeTook / 1e9

    print("Input Size: ", n, "x", n)
    print("Execution time: ", timeTook, " nanoseconds ", "(", timeTookSeconds, " seconds )")

    # Return result and execution time
    return C, end_time - start_time

def main():
    # Test algorithm with three different input sizes (10, 100, 1000)
    # Test for sumArray

    print("Sum Array")
    print("----------")
    inputArray = [i for i in range(10)]
    sumArray(inputArray)
    print("----------")
    inputArray = [i for i in range(100)]
    sumArray(inputArray)
    print("----------")
    inputArray = [i for i in range(1000)]
    sumArray(inputArray)
    print("----------")

    # Test for scalar_multiplication
    print("Scalar Multiplication")
    print("----------")
    scalar_multiplication(10, 5)
    print("----------")
    scalar_multiplication(100, 5)
    print("----------")
    scalar_multiplication(1000, 5)
    print("----------")

    # Test for matrix_multiplication
    print("Matrix Multiplication")
    print("----------")
    n = 10
    A = generate_matrix(n)
    B = generate_matrix(n)
    matrix_multiplication(n, A, B)
    print("----------")
    n = 50
    A = generate_matrix(n)
    B = generate_matrix(n)
    matrix_multiplication(n, A, B)
    print("----------")
    n = 100
    A = generate_matrix(n)
    B = generate_matrix(n)
    matrix_multiplication(n, A, B)
    print("----------")



if __name__ == "__main__":
    main()
